### ISSUE TITLE
Need some help with figuring out my first PSBT

### ISSUE THREAD

**Issue by loxator** [bitcoinjs/bitcoinjs-lib#1771]
Hi there,

I am trying to work with a PSBT and using a test structure which I am getting from another library `@shapeshiftoss/hdwallet-core`. The structure they sent is as follows

```
const txid = "b3002cd9c033f4f3c2ee5a374673d7698b13c7f3525c1ae49a00d2e28e8678ea";
    const hex =  "010000000181f605ead676d8182975c16e7191c21d833972dd0ed50583ce4628254d28b6a3010000008a47304402207f3220930276204c83b1740bae1da18e5a3fa2acad34944ecdc3b361b419e3520220598381bdf8273126e11460a8c720afdbb679233123d2d4e94561f75e9b280ce30141045da61d81456b6d787d576dce817a2d61d7f8cb4623ee669cbe711b0bcff327a3797e3da53a2b4e3e210535076c087c8fb98aef60e42dfeea8388435fc99dca43ffffffff0250ec0e00000000001976a914f7b9e0239571434f0ccfdba6f772a6d23f2cfb1388ac10270000000000001976a9149c9d21f47382762df3ad81391ee0964b28dd951788ac00000000";

    let inputs = [
      {
        addressNList: [0x80000000 + 44, 0x80000000 + 0, 0x80000000 + 0, 0, 0],
        scriptType: "p2pkh",
        amount: String(10000),
        vout: 1,
        txid: txid,
        tx: btcTxJson,
        hex,
      },
    ];

    let outputs = [
      {
        address: "1MJ2tj2ThBE62zXbBYA5ZaN3fdve5CPAz1",
        addressType: "spend",
        scriptType: "p2pkh",
        amount: String(10000 - 1000),
        isChange: false,
      },
    ];

    let res = await wallet.btcSignTx({
      coin: "Bitcoin",
      inputs: inputs,
      outputs: outputs,
      version: 1,
      locktime: 0,
    });

```

I get this structure on my end and try to add the PSBT inputs and outputs and sign them: 

**buildAndSignPSBT**
```
  async buildAndSignPSBT(
    dappMessage: IMessage & { opReturnData?: string },
    walletCredentials: WalletCredentials,
    wallet: IWalletItem
  ): Promise<{ psbt: Bitcoin.Psbt; utxos: UTXO[] }> {
    const { from, recipient } = dappMessage.raw.params[0]
    const {
      coin,
      inputs,
      outputs,
      version,
      locktime,
    } = dappMessage.raw.params?.[0].msg
    if (!walletCredentials) {
      throw new Error(`No credentials passed`)
    }
    const walletData = wallet.walletInfo.data as IHDWalletData
    const phrase = await decryptFromKeystore(
      walletData.keystore as Keystore,
      walletCredentials.walletPassword
    )

    const keyPair = Bitcoin.bip32
      .fromSeed(
        getSeed(phrase),
        btcNetwork(networksToXChainNetwork(dappMessage.network))
      )
      .derivePath(
        this.getFullDerivationPath(
          0,
          networksToXChainNetwork(dappMessage.network)
        )
      )

    const psbt = new Bitcoin.Psbt({ network: btcNetwork(NETWORKS.Mainnet) }) // Network-specific

    psbt.setVersion(version ?? 1)
    locktime && psbt.setLocktime(locktime)
    const btcKeys = this.getBtcKeys(
      phrase,
      walletData.walletIndex,
      Network.Mainnet
    )
    // Build PSBT inputs
    await Promise.all(
      inputs.map(async (input: any) => {
        try {
          const inputData = await this.buildPSBTInputs(input, keyPair)

          psbt.addInput({
            hash: input.txid,
            index: input.vout,
            ...inputData,
          })
        } catch (e) {
          throw new Error(`failed to add input: ${e}`)
        }
      })
    )

    // Build PSBT outputs
    await Promise.all(
      outputs.map(async (output: any) => {
        try {
          const { amount } = output

          let address: string
          if (output.address !== undefined) {
            address = output.address
          } else if (output.addressNList !== undefined) {
            const { publicKey, network } = keyPair
            const payment = this.createPayment(
              publicKey,
              network as Bitcoin.Network,
              output.scriptType
            )
            if (!payment.address)
              throw new Error('could not get payment address')
            address = payment.address
          } else {
            throw new Error('unsupported output type')
          }
          psbt.addOutput({ address, value: Number(amount) })
        } catch (e) {
          throw new Error(`failed to add output: ${e}`)
        }
      })
    )

    if (dappMessage.opReturnData) {
      const data = Buffer.from(dappMessage.opReturnData, 'utf-8')
      const embed = Bitcoin.payments.embed({ data: [data] })
      const script = embed.output
      if (!script) throw new Error('unable to build OP_RETURN script')
      psbt.addOutput({ script, value: 0 })
    }

    psbt.signAllInputs(btcKeys)
    psbt.finalizeAllInputs()

    return { psbt, utxos: [] }
  }
```
**buildPSBTInputs**
```
  async buildPSBTInputs(input: any, keyPair: any) {
    const segwit = ['p2wpkh', 'p2sh-p2wpkh', 'bech32']

    const { amount, hex, scriptType } = input
    const isSegwit = !!scriptType && segwit.includes(scriptType)
    const nonWitnessUtxo = hex && Buffer.from(hex, 'hex')
    const witnessUtxo = input.tx && {
      script: Buffer.from(input.tx.vout[input.vout].scriptPubKey.hex, 'hex'),
      value: Number(amount),
    }
    const utxoData =
      isSegwit && witnessUtxo ? { witnessUtxo } : { nonWitnessUtxo }

    if (!(utxoData.witnessUtxo || utxoData.nonWitnessUtxo)) {
      throw new Error(
        'failed to build input - must provide prev rawTx (segwit input can provide scriptPubKey hex and value instead)'
      )
    }

    const { publicKey, network } = keyPair
    const payment = this.createPayment(publicKey, network, scriptType)

    const scriptData: any = {}
    switch (scriptType) {
      case 'p2sh-p2wpkh':
      case 'p2sh':
      case 'bech32':
        scriptData.redeemScript = payment.redeem?.output
        break
    }

    return {
      ...utxoData,
      ...scriptData,
    }
  }
```
**createPayment**
```
  createPayment(
    pubkey: Buffer,
    network: Bitcoin.Network,
    scriptType?: string
  ): Bitcoin.Payment {
    switch (scriptType) {
      case 'p2sh':
        return Bitcoin.payments.p2sh({ pubkey, network })
      case 'p2pkh':
        return Bitcoin.payments.p2pkh({ pubkey, network })
      case 'p2wpkh':
        return Bitcoin.payments.p2wpkh({ pubkey, network })
      case 'p2sh-p2wpkh':
        return Bitcoin.payments.p2sh({
          redeem: Bitcoin.payments.p2wpkh({ pubkey, network }),
          network,
        })
      case 'bech32':
        return Bitcoin.payments.p2wsh({
          redeem: Bitcoin.payments.p2wsh({ pubkey, network }),
          network,
        })
      default:
        throw new Error(`no implementation for script type: ${scriptType}`)
    }
  }
}

```
I cannot understand what I am doing wrong but I feel like I am signing with the wrong privateKey (which is my key) as I get the message `Cannot sign with keys [keyHere]`

Thank you for your help!

**Comment by junderw** [comment]
1. `addressNList: [0x80000000 + 44, 0x80000000 + 0, 0x80000000 + 0, 0, 0]` is given on the inputs but you're dealing with it on the outputs.

2. You are deriving your seed and path using methods that are opaque to me. Your problem is pubkey mismatch and yet I can't see how you create your pubkeys.

3. `addressNList: [0x80000000 + 44, 0x80000000 + 0, 0x80000000 + 0, 0, 0]` converted into a path for derivePath would be `"m/44'/0'/0'/0/0"` this is the path that much be derived from the root of the BIP32 HD key for your input. (not from some relative path)

**Comment by loxator** [comment]
@junderw , thank you for helping out! Regarding your points

1. The `outputs` could also have an `addressNList` and hence that's the reason for dealing with it there. However, you do have a point that I am not dealing with on the inputs side.

2. The opaque methods that are being used are from [xchainJS](https://github.com/xchainjs/) mainly `decryptFromKeyStore` which return the seedPhrase and then getting a `keyPair` from `bitcoinJS`'s bip32 library.

3. I am not sure how to use this `addressNList` part, should I derive a root path from it which from what I understand will be as you mentioned `"m/44'/0'/0'/0/0"`.  If you could clarify what you meant by `"this is the path that much be derived from the root of the BIP32 HD key for your input. (not from some relative path)"`, I believe it should clear a lot of things up.

Again, thanks a lot for all your help!

**Comment by junderw** [comment]
The return value of `bip32.fromSeed` is "the root of the BIP32 HD key."

Any return value from a call to `derive` or `derivePath` etc. FROM the root would give you a non-root "node" of the HD tree. You could also call this a "relative path." since any further derivation from that node would be relative to that node.

ie.
```
const node = root.derivePath("m/1/2")
const node2 = node.derivePath("3/4")
// node2 has a relative path of 3/4 from node and an absolute path (from root) of m/1/2/3/4
```

The format of `[0x80000000 + 44, 0x80000000 + 0, 0x80000000 + 0, 0, 0]` is very awkward, but it's basically just taking the hardened sign of `'` and converting it to `0x80000000` as per bip32 spec. The BIP32 library does this automatically usually.

You could do something like `addressNList.reduce((node, pathIndex) => node.derive(pathIndex), root)` which would be the same as using derivePath on the result of `addressNList.reduce((path, pi) => path + '/' + (pi & 0x7fffffff) + (pi >= 0x80000000 ? "'" : ''), 'm')`

**Comment by loxator** [comment]
Can I just say you explain really well! As I am still making my way through the world of bitcoin development, your comment about non-root node did fill a few gaps in my knowledge. I have tried to sign the inputs using the node from `addressNList` as follows

```
await Promise.all(
      inputs.map(async (input: any) => {
        try {
          const inputData = await this.buildPSBTInputs(input, root)

          psbt.addInput({
            hash: input.txid,
            index: input.vout,
            ...inputData,
          })

          psbt.signInput(
            0,
            input.addressNList.reduce(
              (node: any, pathIndex: number) => node.derive(pathIndex),
              root
            )
          )
        } catch (e) {
          throw new Error(`failed to add input: ${e}`)
        }
      })
    )
    ```
    
    However, still facing the same issue, i.e. `Cannot sign with key xxxx`, however it shows a different key now. Is what I am doing correct? I am generating the `derivationPath` using the `addressNList` now, using my wallet's root.

**Comment by junderw** [comment]
Your script logic is also weird, especially p2sh and "bech32"

p2sh and p2wsh payments do not have any input for pubkey. And you can't do a p2wsh of a p2wsh.

The only mixed wrapped script that is OK is wrapping a p2wsh inside a p2sh. Any other combo is invalid.

The inner-most wrapped script must contain another non-wrapping payment like p2pkh/p2wpkh etc.

If the script contains p2wsh anywhere you need a witnessScript.

If the script contains p2sh anywhere you need a redeemScript.

If it's a p2sh(p2wsh(xxx)) script, then you need both redeemScript (which is just a p2wsh output) AND the witnessScript (which is the actual script, p2wpkh etc.)

**Comment by loxator** [comment]
Thank you for the helpful pointers! I have amended the function as you have mentioned. However, I am still unable to sign it. I think I am having a difficulty understanding this particular PSBT, as there's already a transaction hash provided, and upon checking the block explorer, it's a transaction made in 2014. 

1. How does me signing it, allow the network to spend my UTXOs? 
2. You had mentioned that the issue is publicKey mismatch. Is there a way to know what publicKey it expects, or did you mean that when I generate the script using the `keyPair` and then try to sign the inputs, there's a mismatch?

Apologize for the super noobish questions, but you have been a great help!

**Comment by loxator** [comment]
Also, do I need a redeemScript / witnessScript in terms of  `p2pkh` transaction?

**Comment by junderw** [comment]
tl;dr

Your online server that is sending you the `addressNList` probably knows all your pubkeys, and is assuming you are using a certain mnemonic seed. (call it X)

If you derive the path from your seed, and it's giving you a pubkey not found error:

1. You haven't added the proper redeemScript/witnessScripts
2. You are using the wrong mnemonic

Are the two most likely scenarios.

Seeing as your script logic (the switch case between scriptTypes) has p2wsh wrapped in p2wsh in its code (which is not valid) the bug is maybe 1... although it might also be 2. Or both, who knows.

**Comment by junderw** [comment]
You're also deriving a single keyPair with bip32 while at the same time calling some magic `this.getBtcKeys()` method and using that to sign all inputs. (which is wrong, since each input could have a different path in addressNList so you need to derive the keypairs for each input individually (most likely))

**Comment by loxator** [comment]
You were right about both the things, just realized they have set up the transaction with a dummy wallet. The inputs were signed now after deriving a path using the `addressNList`, which I wouldn't have gotten without your help. So, thanks a ton!

However, the broadcasted transaction replies back with `bad-txns-inputs-missingorspent`, so I think I am still not building the input correctly. Or maybe don't have enough UTXO in the wallet for the transaction to go through.

I cannot thank you enough for your thorough breakdown of each question and answering them in details, if you have something set up that would allow me to buy you a coffee, please point me to it as that's the least I could do.

**Comment by junderw** [comment]
> the broadcasted transaction replies back with `bad-txns-inputs-missingorspent`

It looks like it was [spent on 2015-08-20 02:56 which is over 6 years ago](https://mempool.space/tx/84da2a3cc29a3e0fb8a3a28882c6fb59a426a95952d484ec2422c5a47b6f93d4) in the 3rd input.

> if you have something set up that would allow me to buy you a coffee, please point me to it as that's the least I could do.

I have Github sponsors active, one time or recurring donations are accepted.

**Comment by loxator** [comment]
Hi again @junderw , back with a super newbie question. Would the PSBT included in bitcoinJS work also for building out the same object in case of Litecoin, in case I have a similar object but with Litecoin addresses? Or is it a completely different thing?

Thought I would use this thread so as to not pollute the issues.

**Comment by junderw** [comment]
Yes, PSBTs work with Litecoin.

In fact, as long as you don't uses addresses and just use script Buffers, there's no need to even include the network object anywhere. It'll just work as is.

**Comment by loxator** [comment]
I see. I have been trying to build and validate a Litecoin PSBT the same way as I have done for Bitcoin. The library throws an error while adding outputs `LLe4PciAJgMMJSAtQQ5nkC13t6SSMmERJ3 has no matching Script `. Any idea? The object is very similar is to what I had posted at the start of this thread 

```
const txid = "1de79c706f34c81bbefad49a9ff8d12b6ca86b77605a1998505e4f8792a5892d";
    const hex =
      "010000000196f5704ef948abb958f32ff216112d3283142baf50723833c378882c14a9adea010000006a47304402207c899ba5197a23b1f3cc4b3621abbc682b5142f3ae29af4b951952573f6c82a002203fd7f038aa8403d2c06fd32c237ab4e915939c25aafa7bcb06fb0ddd46afbfd3012103eddbce765b6d7ae1c91b779696e8b8f72ce444070f83beba2f823af76fd4dfebffffffff0290680a00000000001976a91491e975a0238fa1dfff703e50f062e2544a3e372088aca6791100000000001976a91415757f526dc67b52ae9f74918db532eebc39608688ac00000000";

    const inputs = [
      {
        addressNList: [0x80000000 + 44, 0x80000000 + 2, 0x80000000 + 0, 0, 0],
        scriptType: "spend",
        amount: String(2160258),
        vout: 0,
        txid,
        segwit: false,
        tx: ltcTxJson, //Json contianing tx data
        hex,
      },
    ];

    const outputs = [
      {
        address: "LLe4PciAJgMMJSAtQQ5nkC13t6SSMmERJ3",
        addressType: "spend",
        scriptType: "p2pkh",
        amount: String(261614),
        isChange: false,
      },
    ];
    ```

**Comment by loxator** [comment]
Figured it out, I wasn't passing the property `network` value in the PSBT constructor for Litecoin!
